foundationdb/src/directory/mod.rs

```diff
@@ -0,0 +1,374 @@
+// Copyright 2018 foundationdb-rs developers, https://github.com/Clikengo/foundationdb-rs/graphs/contributors
+// Copyright 2013-2018 Apple, Inc and the FoundationDB project authors.
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Directory provides a tool for managing related subspaces.
+//!
+//! The FoundationDB API provides directories as a tool for managing related Subspaces.
+//! For general guidance on directory usage, see the discussion in the [Developer Guide](https://apple.github.io/foundationdb/developer-guide.html#directories).
+//!
+//! Directories are identified by hierarchical paths analogous to the paths in a Unix-like file system.
+//! A path is represented as a slice of strings. Each directory has an associated subspace used to
+//! store its content. The directory layer maps each path to a short prefix used for the
+//! corresponding subspace. In effect, directories provide a level of indirection for access to subspaces.
+//! Directory operations are transactional.
+//!
+//! It is a direct backport of the [Flow implementation](https://github.com/apple/foundationdb/tree/master/bindings/flow).
+//!
+//! Examples:
+//!
+//! ```rust
+//! use futures::prelude::*;
+//! use foundationdb::directory::Directory;
+//!
+//! async fn async_main() -> foundationdb::FdbResult<()> {
+//!     let db = foundationdb::Database::default()?;
+//!
+//!     // creates a transaction
+//!     let trx = db.create_trx()?;
+//!
+//!     // creates a directory
+//!     let directory = foundationdb::directory::directory_layer::DirectoryLayer::default();
+//!
+//!     // use the directory to create a subspace to use
+//!     let content_subspace = directory.create_or_open(
+//!         // the transaction used to read/write the directory.
+//!         &trx,
+//!         // the path used, which can view as a UNIX path like `/app/my-app`.
+//!         vec![String::from("my-awesome-app"), String::from("my-awesome-user")],
+//!         // do not use any custom prefix or layer
+//!         None, None,
+//!     ).await;
+//!     assert_eq!(true, content_subspace.is_ok());
+//!     
+//!     // Don't forget to commit your transaction to persist the subspace
+//!     trx.commit().await?;
+//!
+//!     Ok(())
+//! }
+//!
+//! // Safe because drop is called before the program exits
+//! let network = unsafe { foundationdb::boot() };
+//! futures::executor::block_on(async_main()).expect("failed to run");
+//! drop(network);
+//! ```
+pub mod directory_layer;
+pub mod directory_partition;
+pub mod directory_subspace;
+pub mod error;
+pub(crate) mod node;
```

Comment:
I think public API hierarchy is too complex, I suggest a "flat" approach.

```suggestion
mod directory_layer;
mod directory_partition;
mod directory_subspace;
mod error;
mod node;
pub use error::DirectoryError;
pub use directory_layer::DirectoryLayer;
pub use directory_partition::DirectoryPartition;
pub use directory_subspace::DirectorySubspace;
```

Reply:
Awesome :star_struck: